format/Stream.fbs

```diff
@@ -0,0 +1,48 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+include "Message.fbs";
+
+namespace org.apache.arrow.flatbuf;
+
+/// ----------------------------------------------------------------------
+/// Arrow stream metadata
+///
+///
+/// The format of the stream is:
+///   - StreamHeader
+///   - Block*
+///
+/// Block:
+///   1. Metadata Length (i32 little endian): Includes 4 byte meta size prefix. Size of 3).
+///   2. Body Length (i32 little endian). Size of 4).
+///   3. RecordBatch header flatbuffer (with i32 size prefix)
+///   4. Data buffers
+/// No padding between the 4 elements.
+
+table StreamHeader {
+  version: org.apache.arrow.flatbuf.MetadataVersion;
+
+  schema: org.apache.arrow.flatbuf.Schema;
+
+  /// Total number of batches in the stream. -1 to indicate unknown.
+  totalBatches: int;
+
+  /// TODO: we can add a flag/length if there are dictionaries.
```

Comment:
I would just say that the stream contains one or more Messages, 

```
table StreamHeader {
  version: org.apache.arrow.flatbuf.MetadataVersion;

  /// Total number of messages in the stream.
  /// * The first message must be the Schema
  /// * Next, 0 or more DictionaryBatch messages
  /// * Then, 1 or more RecordBatch messages
  totalMessages: int;
}
```

The serialized `Message` would need to have a length prefix, then upon reading the Flatbuffer you know how many bytes to read to get the full batch

For evolving dictionaries, we could later allow dictionary batches to appear mid-stream and override existing dictionaries, if that makes sense. 

Sound good? Then I can get cracking on a C++ implementation